### ISSUE TITLE
need help: "Double alias not allowed"

### ISSUE THREAD

**Issue by remdragon** [lark-parser/lark#178]
I have a couple libraries that I have implemented in both Python and PHP because I have to move between the two languages frequently. Keeping feature parity between them is becoming an arduous task. So, I'm creating a language that I call "phpy" which I will use to auto-generate both Python and PHP versions of my libraries. "phpy" will be mostly python 3 syntax with some features removed and disambiguations so that both Python and PHP code can be generated from it without a lot of hacks.

I am using the python3.lark file in lark's examples as my starting point, but I'm getting an error that I can't interpret and neither the documentation nor reading the source code is doing me any good.

The following rule is giving me fits:

```
?atom_expr: ["new"] atom_expr "(" [arguments] ")"  -> funccall
          | atom_expr "[" subscriptlist "]"        -> getitem
          | atom_expr "." NAME                     -> getattr
          | atom
```

I have added ["new"] as an optional keyword at the beginning of funccall so that I can optionally output it depending on whether I'm generating python or php.

I'm getting the following error with that change. I don't understand what a "double alias" is in lark or why my change is attempting to create one.

```
Traceback (most recent call last):
  File "phpy_lark.py", line 30, in <module>
    phpy_parser = lark.Lark.open ( 'phpy.lark', parser='lalr', **kwargs )
  File "C:\Python27\lib\site-packages\lark\lark.py", line 203, in open
    return cls(f, **options)
  File "C:\Python27\lib\site-packages\lark\lark.py", line 157, in __init__
    tokens, self.rules, self.ignore_tokens = self.grammar.compile()
  File "C:\Python27\lib\site-packages\lark\load_grammar.py", line 484, in compile
    expansions = rule_tree_to_text.transform(tree)
  File "C:\Python27\lib\site-packages\lark\visitors.py", line 48, in transform
    return self._transform_tree(tree)
  File "C:\Python27\lib\site-packages\lark\visitors.py", line 44, in _transform_tree
    children = list(self._transform_children(tree.children))
  File "C:\Python27\lib\site-packages\lark\visitors.py", line 39, in _transform_children
    yield self._transform_tree(c) if isinstance(c, Tree) else c
  File "C:\Python27\lib\site-packages\lark\visitors.py", line 45, in _transform_tree
    return self._call_userfunc(tree.data, children, tree.meta)
  File "C:\Python27\lib\site-packages\lark\visitors.py", line 34, in _call_userfunc
    return f(children)
  File "C:\Python27\lib\site-packages\lark\load_grammar.py", line 250, in alias
    assert _alias is None, (alias, expansion, '-', _alias)  # Double alias not allowed
AssertionError: (Token(RULE, u'funccall'), ([NonTerminal(u'atom_expr'), Terminal(Token(TERMINAL, 'LPAR')), Terminal(Token(TERMINAL, 'RPAR'))], None), '-', ([NonTerminal(u'atom_expr'), Terminal(Token(TERMINAL, 'LPAR')), NonTerminal(u'arguments'), Terminal(Token(TERMINAL, 'RPAR'))], None))
```

I have attached the full files involved in case it helps. I intend to open source this project, so have no problem sharing the source code.

[phpy.zip](https://github.com/lark-parser/lark/files/2172244/phpy.zip)

There is very little functionality implemented so far because I am very early in this project.

My background in case it helps: I have written hand-generated recursive descent parsers, a couple lalr bnf parsers, a predictive top-down recursive bnf parser with automatic rule folding ( it's lacking a disambiguation feature that I don't have the time to fix right now ). I have played with ply but it's inability to deal with ambiguity combined with cryptic error messages put a short end to that. I have had success with antlr but strongly dislike java so would prefer to see this project successful because I really like it's design goals.

**Comment by erezsh** [comment]
Hi,

It sounds like an interesting project! 

I think you've stumbled into an unflattering bug in my implementation. The grammar loader compiles

    ["new"] atom_expr "(" [arguments] ")"  -> funccall

Into

    "new" atom_expr "(" [arguments] ")"  -> funccall
    | atom_expr "(" [arguments] ")"  -> funccall

Which (I assume) triggers the safe-guard against defining two rules for the same alias.

Honestly, I'm not sure if this safe-guard is even necessary. And if it is, I need to fix the implementation.

So thanks for bringing this to my attention!

Meanwhile, the following rewrite should work, with the exact same effect:

```ruby
    ?atom_expr: funccall
          | atom_expr "[" subscriptlist "]"        -> getitem
          | atom_expr "." NAME                     -> getattr
          | atom

    funccall: ["new"] atom_expr "(" [arguments] ")" 
```

Good luck, and let me know if you run into any other difficulty.

**Comment by remdragon** [comment]
Thank you, that helps me understand what's happening.

The way I ended up fixing the problem last night was similar ( though I
didn't understand why this fixed the problem ):

?atom_expr: "new" funccall -> newfunccall
          | funccall
          ...

funccall: atom_expr "(" [arguments] ")"

In dealing with this sort of problem in my own latest parser, I drew a
distinction between manually created aliases and automatically generated
ones. Auto-generated aliases always get merged into the parent rule ( so
they end user never receives them in the AST ). You could apply similar
logic here. If you're auto-splitting a rule, flag the duplicate as such and
prevent it from triggering the assert.

A couple counter-points to consider:

1) some users might intentionally want multiple rules to combine into the
same alias because it makes their post-processing logic work better. They
therefore definitely wouldn't want that assert firing or want a mechanism
to tell the engine to allow it in specific circumstances.

An example of where this might occur is a tool that's converting data from
one format to another. Think in terms of converting a proprietary document
to xml. It's not unreasonable to imagine they might want to generate the
same xml tag from multiple rules/aliases in the parse tree. Yes, I know
this example is a bit contrived and that the combining logic might be
easier done in the post-processing logic instead of the grammar, but I'm
trying to look at it from both perspectives and there could be some
practical benefits to this approach.

2) I'm finding I'm moving this logic in a direction where I'm splitting
rules into separate aliases to simplify the post-processing logic. For
example, I have split classdef into 2 separate rules, one with subclassing
and one without. With this change, I get a predictable number of children
of each alias.

Here's what I've done with classdef to illustrate #2:

classdef: "class" NMAE  "(" arguments ")"  ":" suite -> subclassdef
        | "class" NAME ["("           ")"] ":" suite

now, in the implementation for "classdef" I can do something like this:

name, suite = classdef.children
create_class ( name, [], suite )

and for "subclassdef" I can do:

name, bases, suite = subclassdef.children
create_class ( name, bases.to_list(), suite )

I'll probably rewrite funccall to the same effect:

funccall: atom_expr "(" ")"
        | atom_expr "(" arguments ")" -> funccallwithargs

So, in conclusion:

I don't think the assert should go away, but it should give a more
actionable error message, but be suppressed if the duplicate was created by
the engine's auto-refactoring ( because that will only confuse people -
especially newbs to your library like me ).

If it's not too difficult, it might not be a bad idea of having a way to
suppress the alert for specific aliases.

A global flag to disable the assert all-together might be useful in
specific circumstances, but it's use should be discouraged.



Royce Mitchell, IT Consultant
ITAS Solutions
royce3@itas-solutions.com

**Comment by erezsh** [comment]
I pushed a new fix to the `master` branch that should solved this issue. I'll appreciate it if you tried to see if it indeed makes the problem go away.
